File: rust/rbtree-zipper-lambda.rs
use std::rc::Rc;

#[derive(Clone, Copy, PartialEq)]
enum Color { Red, Black }

#[derive(Clone)]
enum Tree {
    Branch(Color, Rc<Tree>, i64, bool, Rc<Tree>),
    Leaf,
}

#[derive(Clone)]
enum Zipper {
    NodeR(Color, Rc<Tree>, i64, bool, Rc<Zipper>),
    NodeL(Color, Rc<Zipper>, i64, bool, Rc<Tree>),
    Done,
}

fn branch(c: Color, l: Rc<Tree>, k: i64, v: bool, r: Rc<Tree>) -> Tree {
    Tree::Branch(c, l, k, v, r)
}

fn rc_t(t: Tree) -> Rc<Tree> { Rc::new(t) }
fn rc_z(z: Zipper) -> Rc<Zipper> { Rc::new(z) }

fn move_up(z: &Zipper, t: Tree) -> Tree {
    match z {
        Zipper::NodeR(c, l, k, v, z1) =>
            move_up(z1, branch(*c, l.clone(), *k, *v, rc_t(t))),
        Zipper::NodeL(c, z1, k, v, r) =>
            move_up(z1, branch(*c, rc_t(t), *k, *v, r.clone())),
        Zipper::Done => t,
    }
}

fn balance_red(z: &Zipper, l: Rc<Tree>, k: i64, v: bool, r: Rc<Tree>) -> Tree {
    match z {
        Zipper::NodeR(Color::Black, l1, k1, v1, z1) =>
            move_up(z1, branch(Color::Black, l1.clone(), *k1, *v1,
                rc_t(branch(Color::Red, l, k, v, r)))),

        Zipper::NodeL(Color::Black, z1, k1, v1, r1) =>
            move_up(z1, branch(Color::Black,
                rc_t(branch(Color::Red, l, k, v, r)),
                *k1, *v1, r1.clone())),

        Zipper::NodeR(Color::Red, l1, k1, v1, z1) => match z1.as_ref() {
            Zipper::NodeR(_, l2, k2, v2, z2) =>
                balance_red(z2,
                    rc_t(branch(Color::Black, l2.clone(), *k2, *v2, l1.clone())),
                    *k1, *v1,
                    rc_t(branch(Color::Black, l, k, v, r))),
            Zipper::NodeL(_, z2, k2, v2, r2) =>
                balance_red(z2,
                    rc_t(branch(Color::Black, l1.clone(), *k1, *v1, l.clone())),
                    k, v,
                    rc_t(branch(Color::Black, r.clone(), *k2, *v2, r2.clone()))),
            Zipper::Done =>
                branch(Color::Black, l1.clone(), *k1, *v1,
                    rc_t(branch(Color::Red, l, k, v, r))),
        },

        Zipper::NodeL(Color::Red, z1, k1, v1, r1) => match z1.as_ref() {
            Zipper::NodeR(_, l2, k2, v2, z2) =>
                balance_red(z2,
                    rc_t(branch(Color::Black, l2.clone(), *k2, *v2, l.clone())),
                    k, v,
                    rc_t(branch(Color::Black, r.clone(), *k1, *v1, r1.clone()))),
            Zipper::NodeL(_, z2, k2, v2, r2) =>
                balance_red(z2,
                    rc_t(branch(Color::Black, l, k, v, r)),
                    *k1, *v1,
                    rc_t(branch(Color::Black, r1.clone(), *k2, *v2, r2.clone()))),
            Zipper::Done =>
                branch(Color::Black,
                    rc_t(branch(Color::Red, l, k, v, r)),
                    *k1, *v1, r1.clone()),
        },

        Zipper::Done => branch(Color::Black, l, k, v, r),
    }
}

fn ins(t: &Tree, k: i64, v: bool, z: Rc<Zipper>, cmp: &dyn Fn(i64, i64) -> i64) -> Tree {
    match t {
        Tree::Branch(c, l, kx, vx, r) => {
            let delta = cmp(k, *kx);
            if delta < 0 {
                ins(l, k, v, rc_z(Zipper::NodeL(*c, z, *kx, *vx, r.clone())), cmp)
            } else if delta > 0 {
                ins(r, k, v, rc_z(Zipper::NodeR(*c, l.clone(), *kx, *vx, z)), cmp)
            } else {
                move_up(&z, branch(*c, l.clone(), *kx, *vx, r.clone()))
            }
        }
        Tree::Leaf => balance_red(&z, rc_t(Tree::Leaf), k, v, rc_t(Tree::Leaf)),
    }
}

fn insert(t: &Tree, k: i64, v: bool) -> Tree {
    ins(t, k, v, rc_z(Zipper::Done), &|a, b| b - a)
}

fn fold(t: &Tree, b: i64) -> i64 {
    match t {
        Tree::Branch(_, l, _k, v, r) => {
            let left = fold(l, b);
            let mid = if *v { left + 1 } else { left };
            fold(r, mid)
        }
        Tree::Leaf => b,
    }
}

fn make_tree(n: i64, t: Tree) -> Tree {
    if n <= 0 {
        t
    } else {
        let m = n - 1;
        let t2 = insert(&t, m, m % 10 == 0);
        make_tree(m, t2)
    }
}

fn fold_test(n: i64) -> i64 {
    let t = make_tree(n, Tree::Leaf);
    fold(&t, 0)
}

fn main() {
    let n: i64 = 4_200_000;
    let p = fold_test(n);
    if p != 420_000 {
        eprintln!("FAIL: expected 420000, got {}", p);
        std::process::exit(1);
    }
}

File: haskell/rbtree-zipper-lambda.hs
{-# LANGUAGE BangPatterns #-}

module Main where

import Data.Int (Int64)

data Color = Red | Black deriving (Eq)

data Tree
  = Branch !Color !Tree !Int64 !Bool !Tree
  | Leaf

data Zipper
  = NodeR !Color !Tree !Int64 !Bool !Zipper
  | NodeL !Color !Zipper !Int64 !Bool !Tree
  | Done

moveUp :: Zipper -> Tree -> Tree
moveUp z t =
  case z of
    NodeR c l k v z1 -> moveUp z1 (Branch c l k v t)
    NodeL c z1 k v r -> moveUp z1 (Branch c t k v r)
    Done -> t

balanceRed :: Zipper -> Tree -> Int64 -> Bool -> Tree -> Tree
balanceRed z l k v r =
  case z of
    NodeR Black l1 k1 v1 z1 ->
      moveUp z1 (Branch Black l1 k1 v1 (Branch Red l k v r))
    NodeL Black z1 k1 v1 r1 ->
      moveUp z1 (Branch Black (Branch Red l k v r) k1 v1 r1)
    NodeR Red l1 k1 v1 z1 ->
      case z1 of
        NodeR _ l2 k2 v2 z2 ->
          balanceRed
            z2
            (Branch Black l2 k2 v2 l1)
            k1
            v1
            (Branch Black l k v r)
        NodeL _ z2 k2 v2 r2 ->
          balanceRed
            z2
            (Branch Black l1 k1 v1 l)
            k
            v
            (Branch Black r k2 v2 r2)
        Done -> Branch Black l1 k1 v1 (Branch Red l k v r)
    NodeL Red z1 k1 v1 r1 ->
      case z1 of
        NodeR _ l2 k2 v2 z2 ->
          balanceRed
            z2
            (Branch Black l2 k2 v2 l)
            k
            v
            (Branch Black r k1 v1 r1)
        NodeL _ z2 k2 v2 r2 ->
          balanceRed
            z2
            (Branch Black l k v r)
            k1
            v1
            (Branch Black r1 k2 v2 r2)
        Done -> Branch Black (Branch Red l k v r) k1 v1 r1
    Done -> Branch Black l k v r

ins :: Tree -> Int64 -> Bool -> Zipper -> (Int64 -> Int64 -> Int64) -> Tree
ins t k v z cmp =
  case t of
    Branch c l kx vx r ->
      let delta = cmp k kx
       in if delta < 0
            then ins l k v (NodeL c z kx vx r) cmp
          else
            if delta > 0
              then ins r k v (NodeR c l kx vx z) cmp
              else moveUp z (Branch c l kx vx r)
    Leaf -> balanceRed z Leaf k v Leaf

insert :: Tree -> Int64 -> Bool -> Tree
insert t k v = ins t k v Done (\a b -> b - a)

makeTree :: Int64 -> Tree -> Tree
makeTree n t
  | n <= 0 = t
  | otherwise =
      let m = n - 1
          t2 = insert t m (m `rem` 10 == 0)
       in makeTree m t2

foldTree :: Tree -> Int -> Int
foldTree t !b =
  case t of
    Branch _ l _ v r ->
      let !left = foldTree l b
          !mid = if v then left + 1 else left
       in foldTree r mid
    Leaf -> b

foldTest :: Int64 -> Int
foldTest n = foldTree (makeTree n Leaf) 0

main :: IO ()
main = do
  let n = 1050000
      p = foldTest n
  if p /= 105000
    then error ("FAIL: expected 105000, got " ++ show p)
    else pure ()
